crates/noirc_evaluator/src/ssa/acir_gen/operations/intrinsics.rs

```diff
@@ -222,3 +235,94 @@ fn prepare_outputs(
     }
     outputs
 }
+
+fn evaluate_println(
+    var_cache: &mut InternalVarCache,
+    memory_map: &mut MemoryMap,
+    is_string_output: bool,
+    args: &[NodeId],
+    ctx: &SsaContext,
+    evaluator: &mut Evaluator,
+) {
+    assert!(args.len() == 1, "print statements can only support one argument");
+    let node_id = args[0];
+
+    let mut log_string = "".to_owned();
+    let mut log_witnesses = Vec::new();
+
+    let obj_type = ctx.object_type(node_id);
+    match obj_type {
+        ObjectType::Pointer(array_id) => {
+            let mem_array = &ctx.mem[array_id];
+            let mut field_elements = Vec::new();
+            for idx in 0..mem_array.len {
+                if let Some(var) = memory_map.load_array_element_constant_index(mem_array, idx) {
+                    let array_elem_expr = var.expression();
+                    if array_elem_expr.is_const() {
+                        field_elements.push(array_elem_expr.q_c);
+                    } else {
+                        let var = match var_cache.get(&node_id) {
+                            Some(var) => var.clone(),
+                            _ => InternalVar::from(array_elem_expr.clone()),
+                        };
+                        if let Some(w) = var.cached_witness() {
+                            log_witnesses.push(*w);
+                        } else {
+                            unreachable!("array element to be logged is missing a witness");
+                        }
+                    }
+                } else {
+                    unreachable!("array element being logged does not exist in memory");
+                }
+            }
+
+            if is_string_output {
+                let final_string = noirc_abi::decode_string_value(&field_elements);
+                log_string.push_str(&final_string);
+            } else if !field_elements.is_empty() {
+                log_string.push('[');
+                let mut iter = field_elements.iter().peekable();
+                while let Some(elem) = iter.next() {
+                    if iter.peek().is_none() {
+                        log_string.push_str(&elem.to_hex());
+                    } else {
+                        log_string.push_str(&format!("{}, ", elem.to_hex()));
+                    }
+                }
+                log_string.push(']');
```

Comment:
Instead of this looping pattern, I like to use `.join`, so something like
```rs
let fields = vecmap(field_elements, |elem| elem.to_hex());
log_string = format!("[{}]", fields.join(", "));
```

Reply:
much cleaner :) done